pkg/kvstore/store/watchstore.go

```diff
@@ -0,0 +1,242 @@
+
+	// NumEntries returns the number of entries synchronized from the store.
+	NumEntries() uint64
+
+	// Drain emits a deletion event for each known key. It shall be called only
+	// when no watch operation is in progress.
+	Drain()
+}
+
+// WatchStoreBackend represents the subset of kvstore.BackendOperations leveraged
+// by WatchStore implementations.
+type WatchStoreBackend interface {
+	// ListAndWatch creates a new watcher for the given prefix after listing the existing keys.
+	ListAndWatch(ctx context.Context, name, prefix string, chanSize int) *kvstore.Watcher
+}
+
+type RWSOpt func(*restartableWatchStore)
+
+// WSWithOnSyncCallback registers a function to be executed after
+// listing all keys from the kvstore for the first time. Multiple
+// callback functions can be registered.
+func RWSWithOnSyncCallback(callback func(ctx context.Context)) RWSOpt {
+	return func(rws *restartableWatchStore) {
+		rws.onSyncCallbacks = append(rws.onSyncCallbacks, callback)
+	}
+}
+
+// WSWithEntriesGauge registers a Prometheus gauge metric that is kept
+// in sync with the number of entries synchronized from the kvstore.
+func RWSWithEntriesMetric(gauge prometheus.Gauge) RWSOpt {
+	return func(rws *restartableWatchStore) {
+		rws.entriesMetric = gauge
+	}
+}
+
+// RSSWithSourceClusterName configures the name of the source cluster the
+// information is synchronized from, which is used to enrich the metrics.
+func RWSWithSourceClusterName(cluster string) RWSOpt {
+	return func(rws *restartableWatchStore) {
+		rws.source = cluster
+	}
+}
+
+type rwsEntry struct {
+	key   Key
+	stale bool
+}
+
+// restartableWatchStore implements the WatchStore interface, supporting
+// multiple executions of the Watch() operation (granted that the previous one
+// already terminated). This allows to transparently handle the case in which
+// we had to create a new etcd connection (for instance following a failure)
+// which refers to the same remote cluster.
+type restartableWatchStore struct {
+	source     string
+	keyCreator KeyCreator
+	observer   Observer
+
+	watching        atomic.Bool
+	synced          bool
+	onSyncCallbacks []func(ctx context.Context)
+
+	// Using a separate entries counter avoids the need for synchronizing the
+	// access to the state map, since the only concurrent reader is represented
+	// by the NumEntries() function.
+	state      map[string]*rwsEntry
+	numEntries atomic.Uint64
+
+	log           *logrus.Entry
+	entriesMetric prometheus.Gauge
+}
+
+// NewRestartableWatchStore returns a WatchStore instance which supports
+// restarting the watch operation multiple times, automatically handling
+// the emission of deletion events for all stale entries (if enabled). It
+// shall be restarted only once the previous Watch execution terminated.
+func NewRestartableWatchStore(keyCreator KeyCreator, observer Observer, opts ...RWSOpt) WatchStore {
+	rws := &restartableWatchStore{
+		source:     option.Config.ClusterName,
```

Comment:
We can pass this with `RWSWithSourceClusterName` right?

Reply:
Yeah, makes sense to me. I've changed it to be mandatory (I'll perform the same change also to the SyncStore, for symmetry).

Reply:
Parallel change for the SyncStore: https://github.com/cilium/cilium/pull/25049/commits/f170cd41c7acf1ce35c4f3ed8ba55743dad00ec4